yang/frr-ospfd.yang

```diff
@@ -0,0 +1,903 @@
+
+	leaf prefix {
+	  type inet:ipv4-prefix;
+	  description
+	    "Area range prefix";
+	}
+	leaf advertise {
+	  type empty;
+	  description
+	    "Advertise this range";
+	}
+	leaf not-advertise {
+	  type empty;
+	  description
+	    "DoNotAdvertise this range";
+	}
+	leaf cost {
+	  type uint32 {
+	    range "0..16777215";
+	  }
+	  description
+	    "Metric for this range";
+	}
+	leaf substitute {
+	  type inet:ipv4-prefix;
+	  description
+	    "Network prefix to be announced instead of range";
+	}
+      }
+    }
+  }
+
+  grouping stub-grouping {
+    container stub {
+      presence "Present when area is strub";
+      
+      leaf no-summary {
+        type empty;
+	  description
+	    "Do not inject inter-area routes into stub";
+      }
+    }
+  }
+
+  grouping shortcut-grouping {
+    container shortcut {
+      leaf default {
+        type empty;
+	description
+	  "Default shortcutting behavior";
+      }
+      leaf disable {
+        type empty;
+	description
+	  "Disable shortcutting through the area";
+      }
+      leaf eanble {
+        type empty;
+	description
+	  "Enable shortcutting through the area";
+      }
+    }
+  }
+
+  grouping authentication-group {
+    container authentication {
+      presence "Enable authentication.";
+      description
+        "Enable authentication on this virtual link.";
+      leaf message-digest {
+        type empty;
+	description
+	  "Use message-digest authentication.";
+      }
+      leaf null {
+        type empty;
+	description
+	  "Use null authentication.";
+      }
+    }
+    list message-digest-key {
+      key "key_id";
+
+      leaf key_id {
+        type uint8;
+	description
+	  "Key id";
+      }
+      leaf mds-key {
+        type string;
+	description
+	  "The OSPF password.";
+      }
+    }
+    leaf authentication-key {
+      type string;
+      description
+        "The OSPF password.";
+    }
+  }
+
+  grouping virtual-link-group {
+    list virtual-link {
+      key "neighbor area_id";
+      description
+        "Virtual link parameters.";
+
+      leaf neighbor {
+        type inet:ipv4-address;
+	description
+	  "Router ID of the remote ABR.";
+      }
+      leaf area_id {
+        type inet:ipv4-address;
+      }
+
+      uses authentication-group;
+
+
+      container timers {
+        leaf dead-interval {
+	  type uint16;
+	  description
+	    "Interval time after which a neighbor is declared down.";
+	}
+	leaf hello-interval {
+	  type uint16;
+	  description
+	    "Time between HELLO packets.";
+	}
+	leaf retransmit-interval {
+	  type uint16;
+	  description
+	    "Time between retransmitting lost link state advertisements.";
+	}
+	leaf transmit-delay {
+	  type uint16;
+	  description
+	    "Link state transmit delay.";
+	}
+      }
+    }
+  }
+
+  grouping area-groupings {
+    container areas {
+      list area {
+        key "area_id";
+	description
+	  "A list of area objects";
+	
+	leaf area_id {
+	  type inet:ipv4-address;
```

Comment:
In the CLI we can specify the OSPF area using either an IPv4 address or an unsigned integer, and the CLI retains the chosen format when displaying the running configuration. With this YANG module we won't be able to preserve that behavior, so we need to discuss if that's acceptable. There are a few options to preserve the old behavior, like using YANG unions, choices, or extending the NB to have the ability to associate metadata to data nodes. But all solutions imply an extra complexity that might not be worth the effort.